kernel: Streamline util library

ryanofsky opened the pull request:
Remove `fees.h`, `errors.h`, and `spanparsing.h` from the util library. Specifically:

- Move `Split` functions from `util/spanparsing.h` to `util/string.h`, using `util` namespace for clarity.
- Move remaining spanparsing functions to `script/parsing.h` since they are used for descriptor and miniscript parsing.
- Combine `util/fees.h` and `util/errors.h` into `common/messages.h` so there is a place for simple functions that generate user messages to live, and these functions are not part of the util library.

Motivation for this change is that the util library is a dependency of the kernel, and we should remove functionality from util that shouldn't be called by kernel code or kernel applications. These changes should also improve code organization and make functions easier to discover. Some of these same moves are (or were) part of #28690, but did not help with code organization, or made it worse, so it is better to move them and clean them up in the same PR so code only has to change one time.

sedited committed:
util: move HexStr and HexDigit from util to crypto

Move HexStr and HexDigit functions from util to crypto. The crypto library does
not actually use these functions, but the consensus library does. The consensus
and util libraries not allowed to depend on each other, but are allowed to
depend on the cryto library, so the crypto library is a reasonable put these.

The consensus library uses HexStr and HexDigit in script.cpp, transaction.cpp,
and uint256.cpp.

The util library does not use HexStr but does use HexDigit in strencodings.cpp
to parse integers.

ryanofsky committed:
util: Move util/string.h functions to util namespace

There are no changes to behavior. Changes in this commit are all additions, and
are easiest to review using "git diff -U0 --word-diff-regex=." options.

Motivation for this change is to keep util functions with really generic names
like "Split" and "Join" out of the global namespace so it is easier to see
where these functions are defined, and so they don't interfere with function
overloading, especially since the util library is a dependency of the kernel
library and intended to be used with external code.

ryanofsky committed:
util: move spanparsing.h Split functions to string.h

This will help move the miniscript / descriptor parsing functions out of the
util library in an upcoming commit, so they are not exposed to libbitcoinkernel
applications. Moving the Split functions should also make them more
discoverable since they now close to related functions like Join.

The functions are moved verbatim without any changes.

ryanofsky committed:
util: move spanparsing.h to script/parsing.h

Move miniscript / descriptor script parsing functions out of util library so
they are not a dependency of the kernel.

There are no changes to code or behavior.

ryanofsky committed:
util: move fees.h and error.h to common/messages.h

Move enum and message formatting functions to a common/messages header where
they should be more discoverable, and also out of the util library, so they
will not be a dependency of the kernel

The are no changes in behavior and no changes to the moved code.

ryanofsky committed:
util: move error.h TransactionError enum to node/types.h

New node/types.h file is analagous to existing wallet/types.h and is a better
place to define simple node types that are shared externally with wallet and
gui code than the util library.

Motivation for this change is to completely remove util/error.h file currently
holding TransactionError in a followup commit.

ryanofsky committed:
util: add TransactionError includes and namespace declarations

Add TransactionError to node namespace and include it directly instead of
relying on indirect include through common/messages.h

This is a followup to a previous commit which moved the TransactionError enum.
These changes were done in a separate followup just to keep the previous commit
more minimal and easy to review.

ryanofsky commented:
Updated 0e647454b49274ee82978c06b606f3fd9c02b779 -> 778c0214d25e5a012c61251d592257ae19805255 ([`pr/rmutil.1`](https://github.com/ryanofsky/bitcoin/commits/pr/rmutil.1) -> [`pr/rmutil.2`](https://github.com/ryanofsky/bitcoin/commits/pr/rmutil.2), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/rmutil.1..pr/rmutil.2)) fixing CI errors from forgitting to add new files to git

re: https://github.com/bitcoin/bitcoin/pull/29015#issuecomment-1844180519

> > Motivation for this change is that the util library is a dependency of the kernel, and we should remove functionality from util that shouldn't be called by kernel code or kernel applications.
> 
> Is the idea here that `util` should be for generic code, and that these bits and pieces (like fees and error) are specific bitcoin features that are only relevant to wallet/p2p, so aren't generic?

I do think that's generally a good thing to do, but it's not really the goal of this PR. The main thing this PR is trying to do is provide a better kernel API and reduce the size of the kernel library.

Because libbitcoin_kernel depends on libbitcoin_util, any external application using the kernel will also depend on util. So we should try to avoid including functions and types in util that are not likely to be useful to external applications, or are unstable and meant for internal use. The common library is usually a better place for these things.

ajtowns commented:
> > > Motivation for this change is that the util library is a dependency of the kernel, and we should remove functionality from util that shouldn't be called by kernel code or kernel applications.
> > 
> > Is the idea here that `util` should be for generic code, and that these bits and pieces (like fees and error) are specific bitcoin features that are only relevant to wallet/p2p, so aren't generic?
> 
> I do think that's generally a good thing to do, but it's not really the goal of this PR. The main thing this PR is trying to do is provide a better kernel API and reduce the size of the kernel library.

Does that mean that things should only go in util if they're going to be used by the kernel library then? That would presumably mean util/bitdeque.h and util/sock.cpp should also eventually go elsewhere, for example? That seems like a bad approach to me (in that it means code has to move into and out of util based on whether it's relevant to the kernel, rather than due to any changes to the code itself, and users then potentially have to deal with it moving into/outof the util:: namespace).

ryanofsky commented:
Updated 778c0214d25e5a012c61251d592257ae19805255 -> 8b21f41335dc837b36ea863156605d092f47a0ab ([`pr/rmutil.2`](https://github.com/ryanofsky/bitcoin/commits/pr/rmutil.2) -> [`pr/rmutil.3`](https://github.com/ryanofsky/bitcoin/commits/pr/rmutil.3), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/rmutil.2..pr/rmutil.3)) to fix clang-tidy error https://cirrus-ci.com/task/4602567305461760. Also switched to `using` declarations more places to avoid creating conflicts and make the PR easier to review

hebasto commented:
@ryanofsky 

https://github.com/bitcoin/bitcoin/pull/29015#issue-2029493410:
> the util library is a dependency of the kernel, and we should remove functionality from util that shouldn't be called by kernel code or kernel applications.

https://github.com/bitcoin/bitcoin/pull/29015#issuecomment-1846332843:
> only code that _shouldn't_ be used by the kernel or kernel applications should be moved out of util, not just code that _isn't_ used by the kernel.

https://github.com/bitcoin/bitcoin/pull/27989#issuecomment-1613966139:
> I think util/ is a good home for general purpose code providing basic data structures and cross platform capabilities, and I think common/ is good home for project-specific code that needs to be shared between bitcoind, bitcoin-cli, and bitcoin-wallet.

Does it make sense to improve https://github.com/bitcoin/bitcoin/blob/master/doc/design/libraries.md by expressing the idea from the quotes above more clearly? As for now, it reads:
> In general, if it is ever unclear whether it is better to add code to util or common, it is probably better to add it to common unless it is very generically useful or useful particularly to include in the kernel.

TheCharlatan commented:
Re https://github.com/bitcoin/bitcoin/pull/29015#issuecomment-1848935768

> Does it make sense to improve https://github.com/bitcoin/bitcoin/blob/master/doc/design/libraries.md by expressing the idea from the quotes above more clearly? As for now, it reads:

This seems like the right place to talk about the relationship between the util and kernel library. My work on the kernel library has been focusing on reducing its essence to validation code, pruning external dependencies, and stripping out as much system related functionality as possible. I think if the goal is to have the kernel be capable of running on as many platforms as possible, potentially beyond the number of platforms that Bitcoin Core is capable of today and extending into embedded environments, this last point becomes crucial.

Looking at the content of `util` it provides roughly two types of functionality: Utilities for basic data structures and their serialization, and cross-platform system related utilities, like file handling, threading, process handling, etc. 

In my opinion, making the `util` library a proper dependency of the kernel should not introduce new cross-platform system-related content into the kernel library. Besides potentially limiting use cases of the kernel, providing cross-platform utilities beyond what the kernel strictly requires seems like a potentially unreasonable future maintenance burden. This would be contrary to our incremental process of creating a minimal kernel library.

The asterisk here is that some of the utilities that are not currently used by the kernel do not introduce new dependencies, since they entirely rely on code that is already used by the kernel. I also generally think that some utilities could remain in the util library even though they are not currently used by validation code. The criteria I propose for evaluating if modules should remain in util after this PR are:

1. The module is directly used by kernel code.
2. The module contains basic data structure utilities for interacting with kernel code.
3. The module does not contain system-related utilities that are not directly used by the kernel.
4. The module does not contain project-specific code that is not relevant to the kernel.
5. The module does not introduce new dependencies beyond the current dependencies of the kernel.

Looking at @hebasto's comment [here](https://github.com/bitcoin/bitcoin/pull/28690#pullrequestreview-1774033780), `threadinterrupt.cpp` and `readwritefile.cpp` would violate the 3rd criteria in my eyes, while the `spanparsing` and `bytevectorhash` modules could potentially remain in `util`.

Between this PR and #28690 I think it would be nice if common criteria for `util` would be defined and documented. All this said, I think it would be acceptable too if things were kept purposefully vague as they are now until the need for more concrete structures arises.

ajtowns commented:
Would it make sense to just merge common into util; but have a "stripped" version of util available for the kernel, that excludes stuff that doesn't match the 5 points above? That way the difference is just at the build system level, it doesn't involve moving files around or moving code between namespaces...

Then the advice would be: (a) just put things in util, (b) but only add things to the "util-core" section of the build file when they're needed by the kernel; (c) keep things in the util:: namespace; (d) split stuff into different .h/.cpp files pretty aggressively?

That would imply a different approach for the [util: move fees.h and error.h to common/messages.h](https://github.com/bitcoin/bitcoin/pull/29015/commits/20b7041f63b1cb0e69347321e20d927e2615c980) commit here.

ryanofsky commented:
re: https://github.com/bitcoin/bitcoin/pull/29015#issuecomment-1849208455

Idea to build util and util-core (or util-lite?) libraries is interesting, because then it would be easy to add/remove things from util library depending on whether the kernel needs them, and it would never require renaming anything. I could see this being useful in the future if util library started growing a lot and accumulating more features of over time.

But for now, IMO, just sticking with the current util/common separation and using the cleanup in this PR would organize things better and be less confusing than having two util versions, and it would require moving fewer things. Even though the libraries.md document is currently ambiguous, I think it should be pretty easy to remove the ambiguity here.

re: https://github.com/bitcoin/bitcoin/pull/29015#issuecomment-1849026409

> The criteria I propose for evaluating if modules should remain in util after this PR are:

I think these are good rules if you replace "basic data structure utilities" with "basic utilities that to do not require outside dependencies", and also drop the 3rd rule. It seems like with the 3rd rule you are trying to make a distinction between "data structure utilities" and "system-related utilities", and I don't think it helps to draw a dividing line there.

ryanofsky committed:
doc: Clarify distinction between util and common libraries in libraries.md

Also describe crypto library which was previously unmentioned

ryanofsky commented:
Added 1 commits b23409474f6fe874a25374bb978db9e5682954e6 -> 6d8cb4d4cd6062ad06cd351439e9139e028995d3 ([`pr/rmutil.5`](https://github.com/ryanofsky/bitcoin/commits/pr/rmutil.5) -> [`pr/rmutil.6`](https://github.com/ryanofsky/bitcoin/commits/pr/rmutil.6), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/rmutil.5...pr/rmutil.6)) updating libraries.md to describe differences between util and common libraries better.

re: https://github.com/bitcoin/bitcoin/pull/29015#issuecomment-1849026409

> My work on the kernel library has been focusing on reducing its essence to validation code, pruning external dependencies, and stripping out as much system related functionality as possible.
> [...]
> In my opinion, making the `util` library a proper dependency of the kernel should not introduce new cross-platform system-related content into the kernel library.

Just to address these points directly, I think the goals of making the kernel portable and working on any platform, and making the kernel minimal working and in constrained environments are great goals. I just don't you need to `rm src/util/sock.cpp` to achieve these goals. As long as the util library is distributed as a static library, or built as a static library and linked into a dynamic library, the socket functions will not be part of resulting executables if they are not called. And if the kernel is being built for a platform which doesn't support sockets at all, `src/util/sock.cpp` can just be stubbed or not compiled at all and there should be no problem for portability either. I do think it's good to keep the util library small and lightweight and remove things which don't fit there. But I don't think it needs to be stripped down in a more extreme way.

TheCharlatan commented:
Re https://github.com/bitcoin/bitcoin/pull/29015#issuecomment-1850626464

> And if the kernel is being built for a platform which doesn't support sockets at all, src/util/sock.cpp can just be stubbed or not compiled at all and there should be no problem for portability either.

One of the guiding principles for the kernel project so far has been to prohibit future re-couplings of decoupled modules. The [tracking issue](https://github.com/bitcoin/bitcoin/issues/27587) says "Any further coupling of our consensus engine with non-consensus modules will result in linker errors, preventing this project from becoming a Sisyphean task of battling coupling regressions.". How could this fit in with leaving unused content in the library?

ryanofsky commented:
> One of the guiding principles for the kernel project so far has been to prohibit future re-couplings of decoupled modules.

I reread the issue you linked to, but I don't understand what it implies here. I do think you can rely on linker errors to ensure kernel code isn't calling wallet code or GUI code or P2P code or storing state in global variables. But what else do you want beyond that? And how does removing `sock.cpp` from the util library help, for example?

ajtowns commented on doc/design/libraries.md:
```diff
@@ -87,14 +88,15 @@ class bitcoin-qt,bitcoind,bitcoin-cli,bitcoin-wallet bold
 - *libbitcoin_consensus* should be a standalone dependency that any library can depend on, and it should not depend on any other libraries itself.
 
-- *libbitcoin_util* should also be a standalone dependency that any library can depend on, and it should not depend on other internal libraries.
+- *libbitcoin_crypto* should be a standalone dependency that any library can depend on, and it should not depend on any other libraries itself. (It is not shown in diagram above to save space, but has the exact same connections as *libbitcoin_consensus*, which is shown.)
```
Doesn't libconsensus necessarily depend on libcrypto (for sha256 at least)?

ajtowns commented on src/node/types.h:
```diff
@@ -0,0 +1,33 @@
+#ifndef BITCOIN_NODE_TYPES_H
+#define BITCOIN_NODE_TYPES_H
+
+enum class TransactionError {
```
It seems odd to put this in "node" when `TransactionErrorString(TransactionError error)` is in "common" ?

`TransactionError` is returned by `CombinePSBTs()` which seems like something that could reasonably be exposed by a command line tool without needing a full node setup.

So putting this in common seems more sensible to me?

ajtowns replied:
> With this background, I thought for this PR putting `TransactionError` in `node` instead of `common` would be better so it would be more likely to be appropriately used for functions like `node::BroadcastTransaction` and `interfaces::Node::broadTransaction`, and less likely to be misused for non-node functions. I could understand the opposite argument, though, that if even if it belongs in `node` when used ideally, it makes more sense for it to be in `common` based on the way it is used now.

I was mostly thinking that it doesn't make sense for `TransactionErrorString()` in common even though it depends on `TransactionError` which is in node, and node depends on common, not vice-versa? I wasn't expecting paragraphs of response :) I mean, it's just an enum with associated translatable descriptions? Doesn't really seem worth worrying about the fact that some of the errors only make sense in a node context; aren't these all just "something went wrong, deal with it, dear user"?

BTW, I think `TransactionError::P2P_DISABLED` can be dropped entirely? (I think the last/only use was removed in #15713)

ajtowns reviewed (commented):
Okay, I think I'm on roughly the same page:

 * keeping Split in util make sense
 * moving our string stuff into the util namespace seems fine * moving the spanparsing stuff to script mostly makes sense -- though I feel a bit like mixing consensus critical script manipulation and miniscript/descriptor parsing in the same namespace is a bit clunky?
 * moving fees.h/error.h to "common" makes sense; though if we can make "common" just be a broader part of "util" that seems better long term.
 * moving TransactionError into node rather than common doesn't make sense to me though

TheCharlatan commented:
Re https://github.com/bitcoin/bitcoin/pull/29015#issuecomment-1850970368

> But what else do you want beyond that? And how does removing sock.cpp from the util library help, for example?

Maybe removing `sock.cpp` is a bad example, since it is already impossible to link against it. `sock.cpp` uses modules that are already not part of the kernel, so trying to link it in will result in errors. Then again, if `sock.cpp` uses common modules, why is it in `util` in the first place? To provide another concrete example,  there is `thread.cpp`, which I was hoping to remove from the kernel in #28960. If it is removed, how could the kernel library be guarded from its future re-introduction if it remains in the `util` library?

ryanofsky reviewed (commented):
Updated 6d8cb4d4cd6062ad06cd351439e9139e028995d3 -> 9ba4dc41fd405a008e9badc4092ef1905f534b96 ([`pr/rmutil.6`](https://github.com/ryanofsky/bitcoin/commits/pr/rmutil.6) -> [`pr/rmutil.7`](https://github.com/ryanofsky/bitcoin/commits/pr/rmutil.7), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/rmutil.6..pr/rmutil.7)) fixing crypto descriptions in libraries.md

re: https://github.com/bitcoin/bitcoin/pull/29015#pullrequestreview-1776320307

>* moving the spanparsing stuff to script mostly makes sense -- though I feel a bit like mixing consensus critical script manipulation and miniscript/descriptor parsing in the same namespace is a bit clunky?

Yes, actually different files in src/script/ directory are built into different libraries. Script interpreting files are part of the consensus library, script generating files are part of the common library, and the script/sigcache file is part of the node library. It might make sense to untangle this in some way.

>* moving fees.h/error.h to "common" makes sense; though if we can make "common" just be a broader part of "util" that seems better long term.

I think I probably agree. It makes sense to me if we are distributing external libraries like consensus & kernel that we should have public utilities which external code can use, and private utilities external code shouldn't use. But names "util" and "common" probably aren't the best way to describe these.

ryanofsky commented:
re: https://github.com/bitcoin/bitcoin/pull/29015#issuecomment-1851506064

> Maybe removing `sock.cpp` is a bad example, since it is already impossible to link against it. `sock.cpp` uses modules that are already not part of the kernel, so trying to link it in will result in errors. Then again, if `sock.cpp` uses common modules, why is it in `util` in the first place?

I'm actually not sure what is this referring to. Running `nm ./src/util/libbitcoin_util_a-sock.o --undefined-only` I see sock.cpp is using LogInstance, NodeClock, and SysErrorString which should all be part of the util library.

> To provide another concrete example, there is `thread.cpp`, which I was hoping to remove from the kernel in #28960. If it is removed, how could the kernel library be guarded from its future re-introduction if it remains in the `util` library?

`thread.h` contains `TraceThread` which seems like a nice utility for the naming threads (which is helpful for monitoring in top) and printing exceptions (which is useful for debugging). I guess kernel code isn't currently using this function (or #28960 removes the last usage), but as long as kernel code is calling threads, what benefit is to preventing kernel code from calling this function? It seems like a useful function, and if anything more code should use it, I would think.

I do agree with goals of making the kernel small and portable, and agree we should organize libraries with that in mind. I also think we should try to trigger link errors to enforce API boundaries. It just seems to me we can do these things without moving a lot around.

TheCharlatan commented:
Re https://github.com/bitcoin/bitcoin/pull/29015#issuecomment-1852688371

> I'm actually not sure what is this referring to. Running nm ./src/util/libbitcoin_util_a-sock.o --undefined-only I see sock.cpp is using LogInstance, NodeClock, and SysErrorString which should all be part of the util library.

Thanks for actually verifying, I just assumed that the `common/system.h`include was actually used :(.

ryanofsky committed:
common: Add PSBTError enum

Add separate PSBTError enum instead of reusing TransactionError enum for PSBT
operations, and drop unused error codes. The error codes returned by PSBT
operations and transaction broadcast functions mostly do not overlap, so using
an unified enum makes it harder to call any of these functions and know which
errors actually need to be handled.

Define PSBTError in the common library because PSBT functionality is
implemented in the common library and used by both the node (for rawtransaction
RPCs) and the wallet.

ryanofsky reviewed (commented):
Updated 9ba4dc41fd405a008e9badc4092ef1905f534b96 -> 414546dcc455fb5f5f8e2c4591b2df1f16863cdc ([`pr/rmutil.7`](https://github.com/ryanofsky/bitcoin/commits/pr/rmutil.7) -> [`pr/rmutil.8`](https://github.com/ryanofsky/bitcoin/commits/pr/rmutil.8), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/rmutil.7..pr/rmutil.8)) adding new commit 18a9f203bba1ad551bdea0d0580b3e07c3715054 to break up TransactionError type and remove PSBT codes
Updated 414546dcc455fb5f5f8e2c4591b2df1f16863cdc -> 7ef33b2bad71a230428b653d6d0297df49630d99 ([`pr/rmutil.8`](https://github.com/ryanofsky/bitcoin/commits/pr/rmutil.8) -> [`pr/rmutil.9`](https://github.com/ryanofsky/bitcoin/commits/pr/rmutil.9), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/rmutil.8..pr/rmutil.9)) to fix PSBT unit test error and clang-tidy errors

TheCharlatan commented:
Concept ACK

hebasto reviewed (commented):
Approach ACK 7ef33b2bad71a230428b653d6d0297df49630d99.

ryanofsky commented on src/node/types.h:
```diff
@@ -17,16 +17,10 @@ enum class TransactionError {
     OK, //!< No error
     MISSING_INPUTS,
     ALREADY_IN_CHAIN,
-    P2P_DISABLED,
```
re: https://github.com/bitcoin/bitcoin/pull/29015#discussion_r1486097145

> In commit [b8900ad](https://github.com/bitcoin/bitcoin/commit/b8900adcd232bcb2b7475be19b5457c91052ab5a): Can you either add a line about its removal in the commit message, or do it in a separate commit?

Thanks, I added a note about dropping unused error codes in the commit message

ryanofsky reviewed (commented):
Thanks for the reviews!

Updated 7ef33b2bad71a230428b653d6d0297df49630d99 -> 486379c10fdce568a64b78cbb2dad1fa52263a79 ([`pr/rmutil.9`](https://github.com/ryanofsky/bitcoin/commits/pr/rmutil.9) -> [`pr/rmutil.10`](https://github.com/ryanofsky/bitcoin/commits/pr/rmutil.10), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/rmutil.9..pr/rmutil.10)) with suggestions

sedited reviewed (approved):
ACK ffcd3068359ea35f3b7c491df866a0d6924e3b37

Regarding the previous discussions on possibly further discriminating the contents of /util, this can always be revisited at a later point in time. Much more important to keep things moving in the correct direction than getting this immediately correct for all possible (and currently unforeseeable) kernel use cases.

sedited reviewed (approved):
Re-ACK 2d9686fa3052ddc52bc129c7ecc8495a5b6c0688

hebasto commented:
I've tested 5eeafeb302dcf56dd5c16c1e4785a41a132344c7 on Ubuntu 23.10.

While all changes look correct and reasonable to me, I was expecting that this PR removes some of dependencies mentioned in https://github.com/bitcoin/bitcoin/issues/28548.

But that is not the case. The `libbitcoin_util` still depends on `libbitcoin_crypto`, `libbitcoin_consensus` and `libbitcoin_common` libraries.

If such an outcome is expected, maybe update the `doc/design/libraries.md` accordingly?

sedited reviewed (approved):
Re-ACK 5eeafeb302dcf56dd5c16c1e4785a41a132344c7

TheCharlatan commented:
Re https://github.com/bitcoin/bitcoin/pull/29015#issuecomment-2083201174

> But that is not the case. The libbitcoin_util still depends on libbitcoin_crypto, libbitcoin_consensus and libbitcoin_common libraries.
> If such an outcome is expected, maybe update the doc/design/libraries.md accordingly?

I understand this PR more as providing clarity of what should be in the util library and #28690 as actually realizing the diagram.

ryanofsky committed:
build: move chainparamsbase from util to common

Move chainparamsbase from util to common, because util library should not
depend on the common library and chainparamsbase uses the ArgsManager class in
common.

ryanofsky committed:
util: move util/message to common/signmessage

Move util/message to common/signmessage so it is named more clearly, and
because the util library is not supposed to depend on other libraries besides
the crypto library. The signmessage functions use CKey, CPubKey, PKHash, and
DecodeDestination functions in the consensus and common libraries.

ryanofsky committed:
build: move memory_cleanse from util to crypto

Move memory_cleanse from util to crypto because the crypto library should not
depend on other libraries, and it calls memory_cleanse.

ryanofsky committed:
test: Add check-deps.sh script to check for unexpected library dependencies

ryanofsky commented:
re: https://github.com/bitcoin/bitcoin/pull/29015#issuecomment-2083201174

> The `libbitcoin_util` still depends on `libbitcoin_crypto`, `libbitcoin_consensus` and `libbitcoin_common` libraries.

Thanks for checking this. Letting the util library depend on the crypto library seems like the easiest path to take because random.cpp, randomenv.cpp, hasher.cpp, and bytevectorhash.cpp are using the crypto hash functions. But I added new commits to the PR to remove the util dependency on crypt, and remove consensus and common dependencies on util. I also added a [script](https://github.com/ryanofsky/bitcoin/blob/pr/rmutil/test/util/check-deps.sh) which turned up a few other other unexpected dependencies (not related to util), which could be addressed separately. I think the script should be able to run in CI, too, but I did not set that up.

---

Rebased 5eeafeb302dcf56dd5c16c1e4785a41a132344c7 -> 6ea5ee5a268cfbb9cd234e0f1381c8a4785ff04c ([`pr/rmutil.13`](https://github.com/ryanofsky/bitcoin/commits/pr/rmutil.13) -> [`pr/rmutil.14`](https://github.com/ryanofsky/bitcoin/commits/pr/rmutil.14), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/rmutil.13-rebase..pr/rmutil.14)) adding dependency checking script and removing unexpected dependencies related to util

hebasto commented:
> I also added a [script](https://github.com/ryanofsky/bitcoin/blob/pr/rmutil/test/util/check-deps.sh) which turned up a few other other unexpected dependencies (not related to util), which could be addressed separately.

It would be great to have such a tool at our disposal.

TheCharlatan commented:
There is an error in commit fd342cb40b63ca00d23946743a038847673f8726, which can be fixed with:
```diff
diff --git a/src/warnings.cpp b/src/warnings.cpp
index 5b6ace63dd..40043cd8e7 100644
--- a/src/warnings.cpp
+++ b/src/warnings.cpp
@@ -11,0 +12 @@
+#include <util/string.h>
```

TheCharlatan commented:
The check script is very useful, could you add it as the first commit? I think it could make review a bit easier.

ryanofsky reviewed (commented):
Updated 6ea5ee5a268cfbb9cd234e0f1381c8a4785ff04c -> 43532d9c9470ad00ecb991859f7f0c6a025fa72e ([`pr/rmutil.14`](https://github.com/ryanofsky/bitcoin/commits/pr/rmutil.14) -> [`pr/rmutil.15`](https://github.com/ryanofsky/bitcoin/commits/pr/rmutil.15), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/rmutil.14..pr/rmutil.15)) fixing silent conflict with #29845, fixing shellcheck lint errors, and moving check-deps.sh commit first as suggested so it is easier to see effects of later commits.

hebasto commented:
There is another CI failure in https://cirrus-ci.com/task/6531594385620992:
```
crypto/.libs/libbitcoin_crypto_base.a(crypto_libbitcoin_crypto_base_la-cleanse.o): in function `memory_cleanse(void*, unsigned long)':
cleanse.cpp:(.text+0x0): multiple definition of `memory_cleanse(void*, unsigned long)'; /usr/bin/ld: DWARF error: invalid or unhandled FORM value: 0x22
support/.libs/libbitcoinkernel_la-cleanse.o:cleanse.cpp:(.text+0x0): first defined here
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

hebasto commented on doc/design/libraries.md:
```diff
@@ -53,21 +54,30 @@ bitcoin-wallet[bitcoin-wallet]-->libbitcoin_wallet_tool;
 libbitcoin_cli-->libbitcoin_util;
 libbitcoin_cli-->libbitcoin_common;
 
+libbitcoin_consensus-->libbitcoin_crypto;
+libbitcoin_consensus-->libbitcoin_crypto;
```
Duplicated lines?

hebasto commented on doc/design/libraries.md:
```diff
@@ -78,22 +88,23 @@ class bitcoin-qt,bitcoind,bitcoin-cli,bitcoin-wallet bold
 </td></tr><tr><td>
 
-**Dependency graph**. Arrows show linker symbol dependencies. *Consensus* lib depends on nothing. *Util* lib is depended on by everything. *Kernel* lib depends only on consensus and util.
+**Dependency graph**. Arrows show linker symbol dependencies. *Crypto* lib depends on nothing. *Util* lib is depended on by everything. *Kernel* lib depends only on consensus and util.
```
```suggestion
**Dependency graph**. Arrows show linker symbol dependencies. *Crypto* lib depends on nothing. *Util* lib is depended on by everything. *Kernel* lib depends only on consensus, crypto and util.
```

hebasto commented on test/util/check-deps.sh:
```diff
@@ -0,0 +1,203 @@
+    "node kernel"
+    "node util"
+    "util crypto"
+    "wallet common"
+    "wallet crypto"
+    "wallet_tool util"
+    "wallet_tool wallet"
+    "wallet util"
```
nit: Sort these lines?

hebasto reviewed (approved):
ACK 850c73e250fe0b248cae80d26561da7047696e3d.

Thanks for the `test/util/check-deps.sh` script!

hebasto reviewed (approved):
re-ACK b58156701ac132f87b8ef8da1c7d22158c804a81, only suggested changes since my recent [review](https://github.com/bitcoin/bitcoin/pull/29015#pullrequestreview-2055510319).

ryanofsky reviewed (commented):
Thanks for the review! Made suggested changes

Updated 850c73e250fe0b248cae80d26561da7047696e3d -> b58156701ac132f87b8ef8da1c7d22158c804a81 ([`pr/rmutil.16`](https://github.com/ryanofsky/bitcoin/commits/pr/rmutil.16) -> [`pr/rmutil.17`](https://github.com/ryanofsky/bitcoin/commits/pr/rmutil.17), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/rmutil.16..pr/rmutil.17)) with suggested changes (documentation fixes, sorting)

sedited reviewed (approved):
Re-ACK b58156701ac132f87b8ef8da1c7d22158c804a81

Good improvements since my last review and having some form of external verification for the work being done is very nice. If other reviewers or maintainers are not happy with introducing a bash script, I could volunteer some time to translate it.

The includes could be cleaned up a bit, especially `util/string.h` and `util/strencodings.h` are missing in a bunch of places. It would also be nice if the newly introduced files had correct includes. Maybe this PR could also be used as a opportunity to more aggressively clean up includes in general.

ryanofsky commented on test/util/check-deps.sh:
```diff
@@ -0,0 +1,203 @@
+#!/usr/bin/env bash
```
re: https://github.com/bitcoin/bitcoin/pull/29015#discussion_r1599843905

> I think rather than `test/` this should go into `contrib/devtools`, which contains functionally similar scripts like `symbol-check`, or `security-check`.

Good suggestion, it does fit in more with those tools, moved over there.

ryanofsky reviewed (commented):
Updated b58156701ac132f87b8ef8da1c7d22158c804a81 -> 7689dfd6646054a0be8e62ccbf72d5403e28b548 ([`pr/rmutil.17`](https://github.com/ryanofsky/bitcoin/commits/pr/rmutil.17) -> [`pr/rmutil.18`](https://github.com/ryanofsky/bitcoin/commits/pr/rmutil.18), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/rmutil.17..pr/rmutil.18)) moving check-deps script.

---

re: https://github.com/bitcoin/bitcoin/pull/29015#pullrequestreview-2055052953

> The includes could be cleaned up a bit, especially `util/string.h` and `util/strencodings.h` are missing in a bunch of places. It would also be nice if the newly introduced files had correct includes. Maybe this PR could also be used as a opportunity to more aggressively clean up includes in general.

Agree this would be nice. The only issue is I never figured out a good way to run the IWYU tool that didn't seem to add lot of unrelated changes to PRs, given that existing includes are so messy, and there are so many that could be cleaned up. If you have any advice on how to run IWYU with this PR, I'd be happy to follow. Or you already have a clear idea of how to clean up includes and just want implement changes on a branch, I'd be happy to use that.

hebasto commented:
There seems to be a conflict with bitcoin/bitcoin#30098. Maybe rebase?

ryanofsky commented:
> There seems to be a conflict with #30098. Maybe rebase?

Thanks! Rebased now.

>  If other reviewers or maintainers are not happy with introducing a bash script, I could volunteer some time to translate it.

Yes this could be a good idea. The script started out just being a few lines of bash listing symbols exported by one library and imported by another one. But then it grew as it was extended to support multiple libraries and suppress known errors. So now it's no longer a small script, though it is pretty clean and well commented. Other possible followups besides translating the script could be:

- Running the script as part of CI
- Cleaning up unexpected dependencies listed as suppressions in the script.
---

Rebased 7689dfd6646054a0be8e62ccbf72d5403e28b548 -> 1c26d10b23bb7a7d0204b4a8b44a50fad89f2a2a ([`pr/rmutil.18`](https://github.com/ryanofsky/bitcoin/commits/pr/rmutil.18) -> [`pr/rmutil.19`](https://github.com/ryanofsky/bitcoin/commits/pr/rmutil.19), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/rmutil.18-rebase..pr/rmutil.19)) due to silent conflict with #30098

sedited reviewed (approved):
Re-ACK 1c26d10b23bb7a7d0204b4a8b44a50fad89f2a2a

hebasto reviewed (approved):
re-ACK 1c26d10b23bb7a7d0204b4a8b44a50fad89f2a2a.

achow101 commented:
CI failure suggests there's a silent merge conflict.

sedited reviewed (approved):
Re-ACK c7376babd19d0c858fef93ebd58338abd530c1f4

hebasto reviewed (approved):
re-ACK c7376babd19d0c858fef93ebd58338abd530c1f4.

My Guix build:
```
x86_64
f33ae90e6d190245d34712f544cd458f282a2c90890029dfbd88d74e8823fb50  guix-build-c7376babd19d/output/aarch64-linux-gnu/SHA256SUMS.part
ee444d54e0b9cebfa6ab0c91fc8b543d414a98338d46d299f3b302273139c81f  guix-build-c7376babd19d/output/aarch64-linux-gnu/bitcoin-c7376babd19d-aarch64-linux-gnu-debug.tar.gz
fdc7400c04fd6bfec5d0ab3148b0f4da110974ecda0859184adad29f38422f6f  guix-build-c7376babd19d/output/aarch64-linux-gnu/bitcoin-c7376babd19d-aarch64-linux-gnu.tar.gz
9ef394d81127cf5ee2eb64508feb0a96f72763073fe0ed9a0901d6008e51b2b8  guix-build-c7376babd19d/output/arm-linux-gnueabihf/SHA256SUMS.part
40cec39c64338a03e7553bb9734d48061d5d1669ab6a59d6250d0dcd5821fde0  guix-build-c7376babd19d/output/arm-linux-gnueabihf/bitcoin-c7376babd19d-arm-linux-gnueabihf-debug.tar.gz
acc8be16daaecab8fdf96763973b8fe7e3e5e9e8014e1e04cf9133cba9f55375  guix-build-c7376babd19d/output/arm-linux-gnueabihf/bitcoin-c7376babd19d-arm-linux-gnueabihf.tar.gz
791971a7afaef6827d3a2b4da5f92a10aa28ff906efd372ef524b900c6815ba7  guix-build-c7376babd19d/output/arm64-apple-darwin/SHA256SUMS.part
bad59a9c07844fa7b788a8b17063c950fa6e124e046d1bee02959344438350dd  guix-build-c7376babd19d/output/arm64-apple-darwin/bitcoin-c7376babd19d-arm64-apple-darwin-unsigned.tar.gz
41de8c1590fc2479ac52946f61affcc592fa7c187abcef02972933e91487ff73  guix-build-c7376babd19d/output/arm64-apple-darwin/bitcoin-c7376babd19d-arm64-apple-darwin-unsigned.zip
b4e834ae5e563d3cf975a0bc5cae1ba9734b19df387eddc45677f910b8bb5043  guix-build-c7376babd19d/output/arm64-apple-darwin/bitcoin-c7376babd19d-arm64-apple-darwin.tar.gz
5d2928dc2f3b6c5390a2020221f1f069bb4c0b0acff09c0d2c79b35ef486b5c9  guix-build-c7376babd19d/output/dist-archive/bitcoin-c7376babd19d.tar.gz
4b8e4f59977136e23ae0dc20816a671257cee4156d80551cbe1c27024857e8f2  guix-build-c7376babd19d/output/powerpc64-linux-gnu/SHA256SUMS.part
d5ce1a3593408c9eb7d1f2fc3d61e7d400381335faeca5cc6be11999c798e4c6  guix-build-c7376babd19d/output/powerpc64-linux-gnu/bitcoin-c7376babd19d-powerpc64-linux-gnu-debug.tar.gz
e11a741a82205faa1c271aa62f5a20329cf46c30db46bda37ed25dc83cbca662  guix-build-c7376babd19d/output/powerpc64-linux-gnu/bitcoin-c7376babd19d-powerpc64-linux-gnu.tar.gz
6ddd58c3fe5e4c600a65b5b9e4bb7e0b99bd3cd5f8655366bb0e29e498e480fe  guix-build-c7376babd19d/output/riscv64-linux-gnu/SHA256SUMS.part
c35ab4f03e23c3889fac7c1a3a64d979af00cb762f6425599149249e8df7e4fd  guix-build-c7376babd19d/output/riscv64-linux-gnu/bitcoin-c7376babd19d-riscv64-linux-gnu-debug.tar.gz
f46bee10789cb0d2f1df0a386b93ebf92ecbd9faaf3446b2df8dc1c5fa77e46c  guix-build-c7376babd19d/output/riscv64-linux-gnu/bitcoin-c7376babd19d-riscv64-linux-gnu.tar.gz
5ba9da03a94da08d34d20edff2d9dd1de729f4087897d3573d5a17fbb053f0d7  guix-build-c7376babd19d/output/x86_64-apple-darwin/SHA256SUMS.part
4b0a6d20c60ff4bfb85bb143b13b0ad28f9c65d536cdbda96b695abf03a0d268  guix-build-c7376babd19d/output/x86_64-apple-darwin/bitcoin-c7376babd19d-x86_64-apple-darwin-unsigned.tar.gz
9525e7082ae7c691381af6abff1a1733b7a2efd772ffbe9bbd98cb522cd8609d  guix-build-c7376babd19d/output/x86_64-apple-darwin/bitcoin-c7376babd19d-x86_64-apple-darwin-unsigned.zip
a62eb1faf7a413267aab397e0c687016a2055f225f5a07d2780eaa7678e0260e  guix-build-c7376babd19d/output/x86_64-apple-darwin/bitcoin-c7376babd19d-x86_64-apple-darwin.tar.gz
70ec027a5d59eefeedf1943c96504d39739753f367442211d0353512b1ac798d  guix-build-c7376babd19d/output/x86_64-linux-gnu/SHA256SUMS.part
793c670bbd8724271c4966cbcffae13a9c876821aeee9ede7449e705acd463e7  guix-build-c7376babd19d/output/x86_64-linux-gnu/bitcoin-c7376babd19d-x86_64-linux-gnu-debug.tar.gz
4c711b6709ee78686f83296bb1e2596772d103cf4fc9959b50085a7d62652921  guix-build-c7376babd19d/output/x86_64-linux-gnu/bitcoin-c7376babd19d-x86_64-linux-gnu.tar.gz
3562a253c9bc448238aad4bb612c73f4fef6899ef297967b4a77b46fab40e73d  guix-build-c7376babd19d/output/x86_64-w64-mingw32/SHA256SUMS.part
3ae8eb90a5dd6fb50caf0f3df16bc0cb60646d1a126a45b89a70bc5d12d00ec4  guix-build-c7376babd19d/output/x86_64-w64-mingw32/bitcoin-c7376babd19d-win64-debug.zip
23f8f28cc9bb4c63d4c4420d8ccf5c0b2547c34b0ebf4b73b10d3553e7fc3681  guix-build-c7376babd19d/output/x86_64-w64-mingw32/bitcoin-c7376babd19d-win64-setup-unsigned.exe
c709e3dbcf3267a6f18ca984b8308bf7f05d58ddbbae83ae9e024db84693a861  guix-build-c7376babd19d/output/x86_64-w64-mingw32/bitcoin-c7376babd19d-win64-unsigned.tar.gz
2cf7011773d95db8398fdb538a5650eb3abd443c2fc18ea091ab07059c778eaf  guix-build-c7376babd19d/output/x86_64-w64-mingw32/bitcoin-c7376babd19d-win64.zip
```

achow101 commented:
ACK c7376babd19d0c858fef93ebd58338abd530c1f4

hebasto commented:
It seems that `libbitcoin_consensus` still depends on `libbitcoin_util` via the `ParseHex()` call in `pubkey.cpp`.

However, the `"consensus util"` dependency is not allowed in `contrib/devtools/check-deps.sh`.

ryanofsky commented:
> It seems that `libbitcoin_consensus` still depends on `libbitcoin_util` via the `ParseHex()` call in `pubkey.cpp`.
> 
> However, the `"consensus util"` dependency is not allowed in `contrib/devtools/check-deps.sh`.

Nice find! This seems to point to a bug in the `check-deps.sh` script for exported template functions. In this case, the fact that pubkey.cpp depends on TryParseHex function is detected in generated `consensus_imports.txt` file:

```
_Z11TryParseHexIhESt8optionalISt6vectorIT_SaIS2_EEESt17basic_string_viewIcSt11char_traitsIcEE libbitcoin_consensus.a:libbitcoin_consensus_a-pubkey.o:
```

And the same symbol is exported according to `nm ./src/util/libbitcoin_util_a-strencodings.o`

```
0000000000000000 W _Z11TryParseHexIhESt8optionalISt6vectorIT_SaIS2_EEESt17basic_string_viewIcSt11char_traitsIcEE
```

but it is a weak symbol, so the script does not detect it. Following fix for the script seems to work:

```diff
--- a/contrib/devtools/check-deps.sh
+++ b/contrib/devtools/check-deps.sh
@@ -78,7 +78,7 @@ extract_symbols() {
     local temp_dir="$1"
     for lib in "${!LIBS[@]}"; do
         for lib_path in ${LIBS[$lib]}; do
-            nm -o "$lib_path" | grep ' T ' | awk '{print $3, $1}' >> "${temp_dir}/${lib}_exports.txt"
+            nm -o "$lib_path" | grep ' T \| W ' | awk '{print $3, $1}' >> "${temp_dir}/${lib}_exports.txt"
             nm -o "$lib_path" | grep ' U ' | awk '{print $3, $1}' >> "${temp_dir}/${lib}_imports.txt"
             awk '{print $1}' "${temp_dir}/${lib}_exports.txt" | sort -u > "${temp_dir}/${lib}_exported_symbols.txt"
             awk '{print $1}' "${temp_dir}/${lib}_imports.txt" | sort -u > "${temp_dir}/${lib}_imported_symbols.txt"
```

Producing the following output:

```
Error: libbitcoin_consensus_a-pubkey.o depends on libbitcoin_util_a-strencodings.o symbol 'std::optional<std::vector<unsigned char, std::allocator<unsigned char> > > TryParseHex<unsigned char>(std::basic_string_view<char, std::char_traits<char> >)', can suppess with:
    SUPPRESS["libbitcoin_consensus_a-pubkey.o libbitcoin_util_a-strencodings.o _Z11TryParseHexIhESt8optionalISt6vectorIT_SaIS2_EEESt17basic_string_viewIcSt11char_traitsIcEE"]=1
Error: Unexpected dependencies were detected. Check previous output.

```

As for the removing the consensus library depenency on util:

https://github.com/bitcoin/bitcoin/blob/a83f050dbe1392fc6b1b6c2a140c7346653b40d3/src/pubkey.cpp#L193

Maybe #30377 could be a good fix for that, since ParseHex is just being used to parse a constant.

hebasto commented:
Ported to the CMake-based build system in https://github.com/hebasto/bitcoin/pull/264.

hebasto commented:
> This is tagged with "Needs CMake port", so if anything needs to be done, now would probably be a good time.

Related to https://github.com/bitcoin/bitcoin/pull/30415.
